Audio playback seek feature

morrolinux opened the pull request:
Since when taking notes we normally write *after* hearing something interesting, just seeking to the exact timestamp a stroke/text was created might just not be enough, so I've added seek forwards and backwards buttons to the playback bar, so that one can make better use of the feature.

morrolinux commented:
Screenshot attached 
![image](https://user-images.githubusercontent.com/20294254/66291738-dd9fc880-e8e2-11e9-8824-e0543a2b37e2.png)

LittleHuba commented:
LGTM

morrolinux commented:
Should I wait until the PR freeze is over before merging? 
What's with the Clang format error? I don't understand what's wrong with the test

LittleHuba commented:
You should most likely wait, just to make the transition smooth.
The Clang Format Error checks the code for right formatting to decrease the merge issues because formats are different. It is currently hard to get this check to signal green as the overall code is not formatted. Part of the PR Freeze is to fromat the code which will then in turn make this check useful. For now you can just ignore it.

LittleHuba commented:
@morrolinux please rebase on master to make this ready for merging.

morrolinux commented:
@LittleHuba 583 files changed... I'm having a bad feeling about this rebase...

LittleHuba commented:
Did you not force-push?

morrolinux commented:
> Did you not force-push?

What I did:
1) rebased my fork's master branch from upstream
2) rebased my feature branch from master
3) fixed conflicts, commit and pushed changes

Now, if I correctly understood what rebasing means, it should be allright, but then, how come that making a PR against upstream master shows all those files changed?
That's the bit I don't understand.

Should I redo it and just try merging my rebased master into my feature branch?

LittleHuba commented:
That is because you did most likely not force-push meaning you created merge commits for all your rebased commits. These are slightly different from the original commits and show therefore on github as changes. Fortunately fixing this is not really hard. You would need to reset your HEAD to before the rebase commits (34e6530), do the rebase locally again and then do the force-push (the force is here very important).

The command should be:
```
git reset --hard 34e6530
git rebase master
# Solve all conflicts
git push --force
```

morrolinux commented:
Alright, `AudioQueue.h` is unfortunately in the PR but as far as I can see the changes should match the master branch, does it look alright to you?

Is it normal for the tests to fail?

LittleHuba commented:
No the tests should definitely not fail. Please check out the test details to review why compilation failed.

LittleHuba commented on src/control/AudioController.cpp:
```diff
@@ -100,6 +100,21 @@ void AudioController::pausePlayback()
 
+void AudioController::seekForwards()
+{
+	XOJ_CHECK_TYPE(AudioController);
```
Remove all XOJ checks as they are no longer valid

LittleHuba commented on src/util/audio/VorbisProducer.cpp:
```diff
@@ -36,23 +38,33 @@ bool VorbisProducer::start(std::string filename, unsigned int timestamp)
 				auto sampleBuffer = new float[1024 * this->sfInfo.channels];
-				std::unique_lock<std::mutex> lock(audioQueue->syncMutex());
-
-				while (!this->stopProducer && numFrames > 0 && !this->audioQueue->hasStreamEnded())
```
You acidentially took the wrong choice in your merge here. You need both the lock and should use `numFrames`.

LittleHuba commented on src/util/audio/VorbisProducer.cpp:
```diff
@@ -36,23 +38,33 @@ bool VorbisProducer::start(std::string filename, unsigned int timestamp)
+						this->seekSeconds = 0;
+					}
+
+					this->audioQueue->push(sampleBuffer, static_cast<unsigned long>(numSamples * this->sfInfo.channels));
```
Use `size_t` in the static cast.

Febbe commented:
If you have problems using the build in merge, you can specify a mergetool like meld. If you call git mergetool it will be opened instead of the console derivate. And you can try to safe your progress, by copiing your branch. By doing that, you'll have infinite tries to do it right.

morrolinux commented:
@Febbe Thanks for the suggestion and @LittleHuba Thanks for the review! I've fixed all of the above, and included the default audio seek setting. 
Everything seems fine, except for the Clang test for which I can't see what's wrong: pressing "Details" button only leads me to a bunch of code with no specific error. I'm not a big expert of Clang formatting myself so how should I interpret this? what to look for?

LittleHuba commented:
You're more then welcome!
The clang check is currently optional since the codebase is not yet formatted and therefore it does not really make sense yet. It's just that we don't get formatting related merge conflicts in the future. Once the whole codebase is formatted PRs can be formatted with a single command before merging. At that point we will make it mandatory.

LittleHuba commented:
By the way the bunch of code you see is a fit diff between your unformatted changes and the correct format.

morrolinux commented:
Aaah, ok, now I get it. I went to the code to fix the formatting and noticed that the codebase itself is indeed not coherent yet. Should I proceed merging then?

LittleHuba commented on ui/settings.glade:
```diff
@@ -3923,6 +3930,71 @@ If available select &lt;small&gt;&lt;tt&gt;pulse&lt;/tt&gt;&lt;/small&gt; as inp
+                                          <object class="GtkLabel" id="label3">
+                                            <property name="visible">True</property>
+                                            <property name="can_focus">False</property>
+                                            <property name="label" translatable="yes">Default Seek Time</property>
```
I'm not really happy with this label. Better would be `Seeking offset in seconds` or along these lines.

LittleHuba replied:
Am I right that this only modifies the offset applied when pressing the seek buttons?

morrolinux replied:
Indeed, that option only modifies the offset applied when pressing the seek buttons.
I thought I'd be more appropriate to keep it separate from the actual track seeking when clicking on a stroke. Maybe I could add another configuration setting for that, but as of now I'm quite happy with it

LittleHuba commented on src/util/audio/VorbisProducer.cpp:
```diff
@@ -32,21 +33,32 @@ bool VorbisProducer::start(std::string filename, unsigned int timestamp)
+					{
+						position += seekSeconds * this->sfInfo.samplerate;
+						if (position < 0) position = 0;
+						sf_seek(this->sfFile, position, SEEK_SET);
```
Using `SEEK_CUR` you can just apply the offset which is then used from the current playback position. That saves you these awkward position calculations all throughout the code.

morrolinux replied:
Damn it really was that simple? I'm gonna try this out when I get home.

LittleHuba replied:
Yeah :joy: 
http://www.mega-nerd.com/libsndfile/api.html#seek

LittleHuba commented on src/util/audio/VorbisProducer.cpp:
```diff
@@ -32,21 +33,27 @@ bool VorbisProducer::start(std::string filename, unsigned int timestamp)
 
 	this->producerThread = new std::thread(
-			[&, filename]
+			[&, filename, seekPosition]
```
This can be removed again as it is now unused in the lambda.

LittleHuba commented on src/util/audio/VorbisProducer.cpp:
```diff
@@ -32,21 +33,27 @@ bool VorbisProducer::start(std::string filename, unsigned int timestamp)
+					if (this->seekSeconds != 0)
+					{
+						sf_seek(this->sfFile,seekSeconds * this->sfInfo.samplerate, SEEK_CUR);
+						this->seekSeconds = 0;
```
This creates a race condition on the variable `seekSeconds` but I think we can live with it as users are unable to click nearly as fast as this poll and reset happens.

morrolinux replied:
Agreed. The worst that can happen is the program skipping a seek command if the user can click twice in a few nanoseconds (which I'm not even sure it's possibile to catch on a non real-time O.S.) so I wouldn't complicate it with synchronization constructs
